website/landing/components/Hero/HeroDesktop/HeroDesktop.tsx

```diff
@@ -0,0 +1,373 @@
+
+  const sectionRef = useRef<HTMLDivElement>(null);
+  const [sectionTop, setSectionTop] = useState(0);
+  const [sectionHeight, setSectionHeight] = useState(0);
+  const [scrollPosition, setScrollPosition] = useState(scrollY.get());
+  const scrollHeight = useMemo(() => sectionHeight / 3, [sectionHeight]);
+  const [animationComplete, setAnimationComplete] = useState(false);
+  const isMounted = useRef(false);
+
+  scrollY.onChange((updatedScroll) => setScrollPosition(updatedScroll));
+
+  useEffect(() => {
+    const isAnimationComplete = scrollPosition >= sectionTop + scrollHeight;
+    if (isAnimationComplete !== animationComplete) {
+      setAnimationComplete(isAnimationComplete);
+    }
+  }, [animationComplete, scrollHeight, scrollPosition, sectionTop]);
+
+  // Push editor into view and adjust wrapper's border radius.
+  const progress = useTransform(
+    scrollY,
+    [sectionTop, sectionTop + scrollHeight],
+    [0, 1]
+  );
+
+  // Subtitle's opacity.
+  const opacity = useTransform(
+    scrollY,
+    [sectionTop + scrollHeight * 0.6, sectionTop + scrollHeight * 0.8],
+    [1, 0]
+  );
+
+  // Push logo pieces into view.
+  const progressInverse = useTransform(
+    scrollY,
+    [sectionTop, sectionTop + scrollHeight],
+    [1, 0]
+  );
+
+  // Rotate logo.
+  const rotate = useTransform(
+    scrollY,
+    [sectionTop + scrollHeight, sectionTop + scrollHeight * 1.2],
+    [-90, 0]
+  );
+
+  // Scale down "fake" preview elements on scroll.
+  const scale = useTransform(
+    scrollY,
+    [sectionTop, sectionTop + scrollHeight],
+    [2.08, 1]
+  );
+
+  // Scale down the whole container on scroll.
+  const containerScale = useTransform(
+    scrollY,
+    [sectionTop, sectionTop + scrollHeight],
+    [1, 0.94]
+  );
+
+  // Sandpack dynamic preview opacity.
+  const sandpackPreviewOpacity = useTransform(
+    scrollY,
+    [
+      sectionTop + (sectionHeight / 4) * 2,
+      sectionTop + (sectionHeight / 4) * 2 + 1,
+    ],
+    [0, 1]
+  );
+
+  // Get dimensions
+  useLayoutEffect(() => {
+    const hero = sectionRef.current;
+    if (!hero) return;
+
+    const onResize = () => {
+      const updatedTop = hero.offsetTop;
+      setSectionTop(updatedTop);
+
+      const updatedHeight = hero.offsetHeight;
+      setSectionHeight(updatedHeight);
+    };
+
+    onResize();
+
+    window.addEventListener("resize", onResize);
+    return () => window.removeEventListener("resize", onResize);
+  }, [sectionRef]);
+
+  useLayoutEffect(() => {
+    if (isMounted.current) return;
+
+    isMounted.current = !!sectionRef.current;
+  }, [sectionRef]);
+
+  return (
+    <SandpackProvider
+      customSetup={{
+        dependencies: { "@stitches/react": "latest" },
+        entry: "./index.js",
+        files,
+      }}
+      template="react"
+    >
+      <AnimatedBox
+        ref={sectionRef}
+        css={{ height: "300vh" }}
+        id="container"
+        style={
+          {
+            "--progress": isMounted ? progress : 0,
```

Comment:
Love this approach

Reply:
it was @marcovincit's idea to define all the variables in a single place, it looks super neat and improves perceived performance 🙌🏽